### ISSUE TITLE
Running Third Generation Sentinel Policy Causes Error

### ISSUE THREAD

**Issue by gist901** [hashicorp/terraform-guides#249]
Hi all,

I downloaded the Sentinel code examples for policies to try running on my local machine. When I attempt to run the example third generation policies for any of the cloud providers I receive the following error message:

```
?[0m?[1m?[31mFAIL - enforce-mandatory-tags.sentinel?[0m
?[0m?[1m?[31m  ERROR - test\enforce-mandatory-tags\fail.json?[0m
?[0m?[31m    Module tfplan-functions installation failed: CreateFile C:\Users\<user>\Desktop\Terraform\Sentinel\MockData\terraform-guides\governance\third-generation\azure\.sentinel\test-modules\C:: The filename, directory name, or volume label syntax is incorrect.
?[0m?[0m?[1m?[31m  ERROR - test\enforce-mandatory-tags\pass.json?[0m
?[0m?[31m    Module tfplan-functions installation failed: CreateFile C:\Users\<user>\Desktop\Terraform\Sentinel\MockData\terraform-guides\governance\third-generation\azure\.sentinel\test-modules\C:: The filename, directory name, or volume label syntax is incorrect.
?[0m
?[0m?[1m?[31mFAIL - restrict-aks-clusters.sentinel?[0m
?[0m?[1m?[31m  ERROR - test\restrict-aks-clusters\fail-max-count.json?[0m
?[0m?[31m    Module tfplan-functions installation failed: CreateFile C:\Users\<user>\Desktop\Terraform\Sentinel\MockData\terraform-guides\governance\third-generation\azure\.sentinel\test-modules\C:: The filename, directory name, or volume label syntax is incorrect.
?[0m?[0m?[1m?[31m  ERROR - test\restrict-aks-clusters\fail-node-count.json?[0m
?[0m?[31m    Module tfplan-functions installation failed: CreateFile C:\Users\<user>\Desktop\Terraform\Sentinel\MockData\terraform-guides\governance\third-generation\azure\.sentinel\test-modules\C:: The filename, directory name, or volume label syntax is incorrect.
?[0m?[0m?[1m?[31m  ERROR - test\restrict-aks-clusters\fail-vm-size.json?[0m
?[0m?[31m    Module tfplan-functions installation failed: CreateFile C:\Users\<user>\Desktop\Terraform\Sentinel\MockData\terraform-guides\governance\third-generation\azure\.sentinel\test-modules\C:: The filename, directory name, or volume label syntax is incorrect.
?[0m?[0m?[1m?[31m  ERROR - test\restrict-aks-clusters\pass-with-autoscaling.json?[0m
?[0m?[31m    Module tfplan-functions installation failed: CreateFile C:\Users\<user>\Desktop\Terraform\Sentinel\MockData\terraform-guides\governance\third-generation\azure\.sentinel\test-modules\C:: The filename, directory name, or volume label syntax is incorrect.
?[0m?[0m?[1m?[31m  ERROR - test\restrict-aks-clusters\pass.json?[0m
?[0m?[31m    Module tfplan-functions installation failed: CreateFile C:\Users\<user>\Desktop\Terraform\Sentinel\MockData\terraform-guides\governance\third-generation\azure\.sentinel\test-modules\C:: The filename, directory name, or volume label syntax is incorrect.
?[0m
?[0m?[1m?[31mFAIL - restrict-app-service-to-https.sentinel?[0m
?[0m?[1m?[31m  ERROR - test\restrict-app-service-to-https\fail.json?[0m
?[0m?[31m    Module tfplan-functions installation failed: CreateFile C:\Users\<user>\Desktop\Terraform\Sentinel\MockData\terraform-guides\governance\third-generation\azure\.sentinel\test-modules\C:: The filename, directory name, or volume label syntax is incorrect.
?[0m?[0m?[1m?[31m  ERROR - test\restrict-app-service-to-https\pass.json?[0m
?[0m?[31m    Module tfplan-functions installation failed: CreateFile C:\Users\<user>\Desktop\Terraform\Sentinel\MockData\terraform-guides\governance\third-generation\azure\.sentinel\test-modules\C:: The filename, directory name, or volume label syntax is incorrect.
?[0m
?[0m?[1m?[31mFAIL - restrict-publishers-of-current-vms.sentinel?[0m
?[0m?[1m?[31m  ERROR - test\restrict-publishers-of-current-vms\fail.json?[0m
?[0m?[31m    Module tfstate-functions installation failed: CreateFile C:\Users\<user>\Desktop\Terraform\Sentinel\MockData\terraform-guides\governance\third-generation\azure\.sentinel\test-modules\C:: The filename, directory name, or volume label syntax is incorrect.
?[0m?[0m?[1m?[31m  ERROR - test\restrict-publishers-of-current-vms\pass.json?[0m
?[0m?[31m    Module tfstate-functions installation failed: CreateFile C:\Users\<user>\Desktop\Terraform\Sentinel\MockData\terraform-guides\governance\third-generation\azure\.sentinel\test-modules\C:: The filename, directory name, or volume label syntax is incorrect.
?[0m
?[0m?[1m?[31mFAIL - restrict-vm-size.sentinel?[0m
?[0m?[1m?[31m  ERROR - test\restrict-vm-size\fail.json?[0m
?[0m?[31m    Module tfplan-functions installation failed: CreateFile C:\Users\<user>\Desktop\Terraform\Sentinel\MockData\terraform-guides\governance\third-generation\azure\.sentinel\test-modules\C:: The filename, directory name, or volume label syntax is incorrect.
?[0m?[0m?[1m?[31m  ERROR - test\restrict-vm-size\pass.json?[0m
?[0m?[31m    Module tfplan-functions installation failed: CreateFile C:\Users\<user>\Desktop\Terraform\Sentinel\MockData\terraform-guides\governance\third-generation\azure\.sentinel\test-modules\C:: The filename, directory name, or volume label syntax is incorrect.
?[0m
```
I can successfully run the second generation policy examples:

```?[0m?[1m?[32mPASS - enforce-mandatory-tags.sentinel?[0m
?[0m?[1m?[32m  PASS - test\enforce-mandatory-tags\fail-missing-tag-0.11.json?[0m
?[0m?[1m?[32m  PASS - test\enforce-mandatory-tags\fail-missing-tag-0.12.json?[0m
?[0m?[1m?[32m  PASS - test\enforce-mandatory-tags\fail-no-tags-0.11.json?[0m
?[0m?[1m?[32m  PASS - test\enforce-mandatory-tags\fail-no-tags-0.12.json?[0m
?[0m?[1m?[32m  PASS - test\enforce-mandatory-tags\pass-0.11.json?[0m
?[0m?[1m?[32m  PASS - test\enforce-mandatory-tags\pass-0.12.json?[0m
?[0m?[1m?[32mPASS - restrict-app-service-to-https.sentinel?[0m
?[0m?[1m?[32m  PASS - test\restrict-app-service-to-https\fail-0.12.json?[0m
?[0m?[1m?[32m  PASS - test\restrict-app-service-to-https\pass-0.12.json?[0m
?[0m?[1m?[32mPASS - restrict-publishers-of-current-vms.sentinel?[0m
?[0m?[1m?[32m  PASS - test\restrict-publishers-of-current-vms\fail-0.11.json?[0m
?[0m?[1m?[32m  PASS - test\restrict-publishers-of-current-vms\fail-0.12.json?[0m
?[0m?[1m?[32m  PASS - test\restrict-publishers-of-current-vms\pass-0.11.json?[0m
?[0m?[1m?[32m  PASS - test\restrict-publishers-of-current-vms\pass-0.12.json?[0m
?[0m?[1m?[32mPASS - restrict-vm-size.sentinel?[0m
?[0m?[1m?[32m  PASS - test\restrict-vm-size\fail-0.11.json?[0m
?[0m?[1m?[32m  PASS - test\restrict-vm-size\fail-0.12.json?[0m
?[0m?[1m?[32m  PASS - test\restrict-vm-size\pass-0.11.json?[0m
?[0m?[1m?[32m  PASS - test\restrict-vm-size\pass-0.12.json?[0m
```

My current version of Sentinel is the following:```Sentinel v0.16.1``` and I have tried running the command using administrator rights. I am running this on a Windows 10 machine.

**Comment by hcrhall** [comment]
@gist901 this issue should now be resolved with the [latest version](https://releases.hashicorp.com/sentinel/0.17.3/) of the Sentinel CLI. We discovered an issue with paths that only occurs when running the CLI Microsoft Windows which we fixed in Sentinel 0.17.3

**Comment by rberlind** [comment]
@gist901 : Were you able to retest this with a newer version of the Sentinel CLI on Windows?
If so, did it work for you?
Can we close this issue?

Roger Berlind

**Comment by SeeObjective** [comment]
This issue still exists with 0.18.4, are there any known workarounds?

**Comment by rberlind** [comment]
@hcrhall : Someone reported having this problem with 0.18.4, presumably on Windows. I received a notification in connection with this issue for it, but do not see any comment on this issue itself.

**Comment by hcrhall** [comment]
> @hcrhall : Someone reported having this problem with 0.18.4, presumably on Windows. I received a notification in connection with this issue for it, but do not see any comment on this issue itself. Maybe they deleted their comment? In any case, can you confirm that the third generation policies should work on Windows with 0.18.4?

@rberlind I was just testing this on Windows 10 Pro (20H2) against the [Azure](https://github.com/hashicorp/terraform-guides/tree/master/governance/third-generation/azure) policies and the tests all passed 🤷

**Comment by rberlind** [comment]
Thanks @hcrhall

**Comment by SeeObjective** [comment]
Hello, sorry about the comment confusion - I wasn't sure whether to report a new issue for this. I get a slightly different file path error when using remote sources (downloading from github) for policies/modules. This error is seen on 0.18.4 for windows and the exact same .hcl file + mocks work on MacOS.
 
`←[31mtfplan-functions installation failed: parse "file://C:\\Users\\test\\AppData\\Local\\Temp\\sentinel-policy102789275\\tfplan-functions.sentinel": invalid port ":\\Users\\test\\AppData\\Local\\Temp\\sentinel-policy102789275\\tfplan-functions.sentinel" after host←[0m`

My understanding is that the policies are being downloaded and stored on my local machine, but the file:// is being interpreted as a host+port combination instead of the file protocol - not sure why.

**Comment by rberlind** [comment]
@SeeObjective : To better help you, we would need to know whether you are seeing this issue with the Sentinel CLI or with Terraform Cloud or Terraform Enterprise. Since you mentioned mocks, I think you are using the Sentinel CLI. If using TFC or TFE, we would also need to see your sentinel.hcl policy set definition file so we can see exactly how you are referencing the remote module.  If using the Sentinel CLI, please share your configuration files and test cases used with the `sentinel apply` or `sentinel test` command that gives the error and indicate which command gives the error.  Thanks.

**Comment by SeeObjective** [comment]
Thanks @rberlind! I'm using the CLI on windows, and while I have a longer .hcl file, I'm able to reproduce this issue with a much simpler file: 

The sentinel.hcl contains

```
policy "azure-cis-3.3-storage-queue-logging-is-enabled" {
  source            = "git::https://github.com/hashicorp/terraform-foundational-policies-library//cis/azure/storage/azure-cis-3.3-storage-queue-logging-is-enabled/azure-cis-3.3-storage-queue-logging-is-enabled.sentinel"
  enforcement_level = "advisory"
}
```
and the command I execute is 

`sentinel apply azure-cis-3.3-storage-queue-logging-is-enabled` 
 
which errors out like this - 

```
C:\Users\test\test\sentinel-checks - Copy>sentinel apply azure-cis-3.3-storage-queue-logging-is-enabled
←[0m←[1m←[31mPolicy installation failed←[0m

←[31mazure-cis-3.3-storage-queue-logging-is-enabled installation failed: parse "file://C:\\Users\\test\\AppData\\Local\\Temp\\sentinel-policy209698643\\azure-cis-3.3-storage-queue-logging-is-enabled.sentinel": invalid port ":\\Users\\test\\AppData\\Local\\Temp\\sentinel-policy209698643\\azure-cis-3.3-storage-queue-logging-is-enabled.sentinel" after host←[0m
```

The check runs fine on MacOS.

**Comment by rberlind** [comment]
@SeeObjective : I was able to replicate the problem on Windows 7 with the 64-bit executable of Sentinel 0.18.4.
I think the problem is that git clones the remote policy into a directory and then uses a path "c:\\Users\\test\\AppData\\Local\\Temp\\sentinel-policy209698643\\azure-cis-3.3-storage-queue-logging-is-enabled.sentinel" but interprets the portion after the : as a port instead of as a path.

I should note a few things:

1. To use your sentinel.hcl on MacOS, I had to change the name of the policy from "azure-cis-3.3-storage-queue-logging-is-enabled" to "azure-cis-3-3-storage-queue-logging-is-enabled" and then run `sentinel apply azure-cis-3-3-storage-queue-logging-is-enabled`.  Otherwise, I got an error "Error reading policy: open azure-cis-3.3-storage-queue-logging-is-enabled: no such file or directory".
2. Also, I had to add a mock file under a new mocks directory under the directory with sentinel.hcl and then create a file in it called "tfplan-fail.sentinel" which I populated with the contents of https://github.com/hashicorp/terraform-foundational-policies-library/blob/master/cis/azure/storage/azure-cis-3.3-storage-queue-logging-is-enabled/testdata/mock-tfplan-failure.sentinel.
3. I then added the following to sentinel.hcl:
```
mock "tfplan/v2" {
  module {
    source = "mocks/tfplan-fail.sentinel"
  }
}
```

Having established that I could run the new sentinel.hcl file and mock on MacOS, I then zipped up my directory and copied it to my Windows 7 laptop.

I actually could not get it to run correctly on Windows 7 due to issues with the length of some of the file names.  But I then created a new test case that used the remote policy https://github.com/hashicorp/terraform-guides/blob/master/governance/third-generation/aws/restrict-ec2-instance-type.sentinel and the mock file https://github.com/hashicorp/terraform-guides/blob/master/governance/third-generation/aws/test/restrict-ec2-instance-type/mock-tfplan-fail.sentinel (renamed to "mocks/tfplan-fail.sentinel".

After doing that and running `sentinel apply restrict-ec2-instance-type.sentinel` I was able to get an error like yours.

I should note that I ran my test in a standard Window cmd shell, not powershell or some other more advanced shell.

So, I do think there is still an issue with running the Sentinel CLI with some remote sources on Windows (at least Windows 7).

However, I'm going to close this issue because the problem is not caused by anything in this repository.  Instead, I am going to open an internal Asana issue requsting that the Sentinel team investigate this issue further.

**Comment by rberlind** [comment]
One workaround might be to use a raw GitHub URL in the source instead of the `git::` option.  I will test this.

**Comment by rberlind** [comment]
@gist901 and @SeeObjective : I **was** able to workaround the issue by using raw GitHub URLs instead of the `git::` URLs.
So, for the foundational policy, I used "https://raw.githubusercontent.com/hashicorp/terraform-foundational-policies-library/master/cis/azure/storage/azure-cis-3.3-storage-queue-logging-is-enabled/azure-cis-3.3-storage-queue-logging-is-enabled.sentinel" as the source.  But to get this to work, I did have to change "3.3" in the name of the policy to "3-3" and then run `sentinel apply` with "3-3" instead of "3.3". I don't think Sentinel likes `.` in name of a policy except for just before the `sentinel` extension, at least not on windows.

So, my sentinel.hcl ended up being:

```
policy "azure-cis-3-3-storage-queue-logging-is-enabled" {
  source = "https://raw.githubusercontent.com/hashicorp/terraform-foundational-policies-library/master/cis/azure/storage/azure-cis-3.3-storage-queue-logging-is-enabled/azure-cis-3.3-storage-queue-logging-is-enabled.sentinel"
  enforcement_level = "advisory"
}

mock "tfplan/v2" {
  module {
    source = "mocks/tfplan-fail.sentinel"
  }
}
```

And I then ran `sentinel apply azure-cis-3-3-storage-queue-logging-is-enabled`.
